kube/src/api/streaming.rs

```diff
@@ -0,0 +1,274 @@
+// Code to support streaming resources (attach/exec/portforward)
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    sync::Mutex,
+    task::{Context, Poll, Waker},
+};
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::Status;
+
+use async_tungstenite::{
+    tokio::ConnectStream,
+    tungstenite::{self as ws, Message},
+    WebSocketStream,
+};
+use futures::{future::Either, SinkExt, Stream, StreamExt, TryStreamExt};
+use futures_util::future::select;
+use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt, DuplexStream};
+use tokio_util::codec;
+
+// Internal state of an attached process
+struct AttachedProcessState {
+    waker: Option<Waker>,
+    finished: bool,
+    status: Option<Status>,
+    stdin_writer: Option<DuplexStream>,
+    stdout_reader: Option<DuplexStream>,
+    stderr_reader: Option<DuplexStream>,
```

Comment:
These are `DuplexStream` (because it's created by `tokio::io::duplex()`), but this fact is hidden from outside and behave as simplex. See associated methods below.